### DISCUSSION TITLE
Architecture description between zigbee2mqtt, herdsman, and herdsman-converters

### DISCUSSION THREAD

**Discussion by junosuarez** [Koenkk/zigbee2mqtt#17172] [General]
I'm wanting to understand deeper the design and implementation of zigbee2mqtt. Is there a description of the relation between  zigbee2mqtt, herdsman, and herdsman-converters? (and maybe also the webui repo). What types of things belong in which repo? Are they released independently or together in lockstep? 

As I've been exploring on my own, I've been surprised several times - most recently, seeing that a significant part of the Groups and Scenes logic is implemented in zigbee-herdsman-converters `toZigbee.js`.  

Any other resources aimed at developing zigbee2mqtt (key github issues/discussions, blog posts, etc), much appreciated!

**Comment by Koenkk** [comment]
- Zigbee-herdsman: used by Zigbee2MQTT, this library is responsible for communicating with the adapter, managing a database of connected devices, interviewing devices upon joining, implementing the ZCL, etc. Zigbee-herdsman is a fork of zigbee-shepherd (https://github.com/zigbeer/zigbee-shepherd)
- Zigbee-herdsman-converters: this project contains the device drivers. It converts Zigbee messages (in ZCL format) send by the devices (`fromZigbee`) to more friendly ones, e.g. an object `{"temperature": 20}`. It also contains the knowledge on how to control devices (`toZigbee`), e.g. what Zigbee command should be send on `{"state": "ON"}`. This project was initially in Zigbee2MQTT and was put in a separate project such that other projects (like ioBroker.zigbee) could re-use it. 
- Zigbee2MQTT: the glue between MQTT/zigbee-herdsman/zigbee-herdsman-converters. Contains mainly the logic to convert from and to MQTT.

As you've already figured out, a feature can usually be implemented in multiple projects (with both pros and cons). Take for example the availability feature, this could also have been implemented in Zigbee-herdsman but is in Zigbee2MQTT. 

In terms of releasing, most of the times releases are independent but of course Zigbee-hersman API changes also influence Zigbee-herdsman-converters, then they have to be released together.

**Comment by junosuarez** [comment]
Awesome, thanks for taking the time to describe that!

1. I've noticed there's a bot to automatically pull in updates from zigbee-herdsman and zigbee-herdsman-converters into zigbee2mqtt. Would it make sense to have these separate packages in a monorepo to reduce this friction? It makes sense to me what you've described to have them as separate packages for modularity and re-use. However, I've noticed most issues for new device support, etc come into this repo as the "main" project repo, even though the implementation often involves those other repos. It makes it harder to link code changes to issues when they are spread across separate repos. I know this might take a good amount of effort, but I'm wondering if you've considered or would be open to this change.

2. zigbee2mqtt is in typescript, while zigbee-herdsman and zigbee-herdsman-converters are in plain javascript. From the commit history, it seems like the conversion to typescript for z2m happened relatively recently. Ultimately, do you see zigbee-herdsman and zigbee-herdsman-converters migrating to typescript as well?

**Reply by Koenkk** [reply]
1. I don't experience any friction here (or maybe I don't understand what you mean?), the whole releasing and updating procedure is fully automated.
2. Zigbee-herdsman is also in typescript (before z2m went Typescript). There was a discussion wether zigbee-herdsman-converters should be migrated to ts, you can find it here: https://github.com/Koenkk/zigbee-herdsman-converters/pull/5322
